python/lsst/meas/algorithms/stamps.py

```diff
@@ -0,0 +1,351 @@
+        llcY = options["llcY"]
+        width = options["width"]
+        height = options["height"]
+        bbox = Box2I(Point2I(llcX, llcY), Extent2I(width, height))
+        kwargs["bbox"] = bbox
+    stamp_parts = {}
+    idx = 1
+    while len(stamp_parts) < nStamps:
+        md = afwFits.readMetadata(filename, hdu=idx)
+        if md['EXTNAME'] in ('IMAGE', 'VARIANCE'):
+            reader = afwImage.ImageFitsReader(filename, hdu=idx)
+        elif md['EXTNAME'] == 'MASK':
+            reader = afwImage.MaskFitsReader(filename, hdu=idx)
+        else:
+            raise ValueError(f"Unknown extension type: {md['EXTNAME']}")
+        stamp_parts.setdefault(md['EXTVER'], {})[md['EXTNAME'].lower()] = reader.read(**kwargs)
+        idx += 1
+    # construct stamps themselves
+    stamps = []
+    meta_dict = metadata.toDict()
+    for k in stamp_parts:
+        maskedImage = afwImage.MaskedImageF(**stamp_parts[k])
+        # Assume the value of EXTVER is the index into the metadata
+        stamps.append(stamp_factory(maskedImage, meta_dict, k))
+
+    return stamps, metadata
+
+
+@dataclass
+class Stamp:
+    """Single stamp
+
+    Parameters
+    ----------
+    stamp_im : `lsst.afw.image.MaskedImageF`
+        The actual pixel values for the postage stamp
+    position : `lsst.geom.SpherePoint`
+        Position of the center of the stamp.  Note the user
+        must keep track of the coordinate system
+    size : `int`
+        The size of the stamp in pixels
+    """
+    stamp_im: afwImage.maskedImage.MaskedImageF
+    position: SpherePoint
+    size: int
+
+    @classmethod
+    def factory(cls, stamp_im, metadata, index):
+        """This method is needed to service the FITS reader.
+        We need a standard interface to construct objects like this.
+        Parameters needed to construct this object are passed in via
+        a metadata dictionary and then passed to the constructor of
+        this class.  If lists of values are passed with the following
+        keys, they will be passed to the constructor, otherwise dummy
+        values will be passed: RA_DEG, DEC_DEG, SIZE.  They should
+        each point to lists of values.
+
+        Parameters
+        ----------
+        stamp : `lsst.afw.image.MaskedImage`
+            Pixel data to pass to the constructor
+        metadata : `dict`
+            Dictionary containing the information
+            needed by the constructor.
+        idx : `int`
+            Index into the lists in ``metadata``
+
+        Returns
+        -------
+        stamp : `Stamp`
+            An instance of this class
+        """
+        if 'RA_DEG' in metadata and 'DEC_DEG' in metadata and 'SIZE' in metadata:
+            return cls(stamp_im=stamp_im,
+                       position=SpherePoint(Angle(metadata['RA_DEG'][index], degrees),
+                                            Angle(metadata['DEC_DEG'][index], degrees)),
+                       size=metadata['SIZE'][index])
```

Comment:
Why can't this be worked out from the masked image?

Reply:
Maybe I should assert that the size of the postage stamp is the same as the size that was asked for?

Reply:
I'm not sure why size needs to be specified in the constructor. Maybe I've misunderstood that metadata field.

Checking that the metadata is consistent with the masked image seems reasonable, but if they are meant to be the same then there is no need to pass it in to the constructor.

Reply:
My understanding is that is how `dataclass` objects work, maybe I can give it a default so I can omit it from the constructor

Reply:
You are saying take size out of the dataclass entirely since it is an attribute of the masked image anyway.  I see your point.

Reply:
Yes, I am asking why you need to set it when you know it already. Using SIZE to check consistency during construction is fine but I don't understand why it needs to be part of the dataclass.

Reply:
OK.  I'm leaving it out, then.

Reply:
You can of course add a `getSize()` method if the size is needed, and that can query the masked image.